File: src/ray.rs
use crate::vec3::{Point3, Vec3};

pub(crate) struct Ray {
    origin: Point3,
    direction: Vec3,
}

impl Default for Ray {
    fn default() -> Self {
        Self {
            origin: Point3::default(),
            direction: Vec3::default(),
        }
    }
}

impl Ray {
    pub(crate) fn new(origin: Point3, direction: Vec3) -> Self {
        Self { origin, direction }
    }

    pub(crate) fn origin(&self) -> &Point3 {
        &self.origin
    }

    pub(crate) fn direction(&self) -> &Vec3 {
        &self.direction
    }

    pub(crate) fn at(&self, t: f64) -> Point3 {
        self.origin + t * self.direction
    }
}

File: src/rtweekendpkg/ray.nim
import vec3

type Ray* = object
  orig: Point3
  dir: Vec3

proc newRay*(origin: Point3, dir: Vec3): Ray =
  Ray(orig: origin, dir: dir)

proc newRay*: Ray =
  newRay(newPoint3(), newVec3())

proc origin*(r: Ray): Point3 =
  r.orig

proc direction*(r: Ray): Vec3 =
  r.dir

proc at*(r: Ray, t: float): Point3 =
  r.orig + t * r.dir
